crates/matrix-sdk-ui/src/encryption_sync/mod.rs

```diff
@@ -34,8 +34,27 @@ use futures_util::{pin_mut, StreamExt};
 use matrix_sdk::{Client, SlidingSync};
 use matrix_sdk_crypto::store::locks::CryptoStoreLock;
 use ruma::{api::client::sync::sync_events::v4, assign};
+use tokio::sync::OwnedMutexGuard;
 use tracing::{debug, trace};
 
+/// Unit type representing a permit to *use* an [`EncryptionSync`].
+///
+/// This must be created once in the whole application's lifetime, wrapped in a
+/// mutex. Using an `EncryptionSync` must then lock that mutex in an owned
+/// way, so that there's at most a single `EncryptionSync` running at any time
+/// in the entire app.
+pub struct EncryptionSyncPermit(());
+
+impl EncryptionSyncPermit {
+    /// Create a new [`EncryptionSyncPermit`].
+    ///
+    /// Note: in general, you'd want to get such a permit from a [`SyncService`]
+    /// instead of creating it yourself.
+    pub fn new() -> Self {
```

Comment:
It is questionable to make `new` public. Maybe `pub(super)` or `pub(crate)`? Making it public is a weakness in this design I believe, as it breaks the uniqueness of this permit.